File: MaxScoreBYREmovingSubstring.py
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        def remove_substring(input: str, target_pair: str) -> (str, int):
            char_stack, count = [], 0
            for char in input:
                if char_stack and char_stack[-1] == target_pair[0] and char == target_pair[1]:
                    char_stack.pop()
                    count += 1
                else:
                    char_stack.append(char)
            return "".join(char_stack), count

        high_priority_pair, low_priority_pair = ("ab", "ba") if x > y else ("ba", "ab")
        s, high_count = remove_substring(s, high_priority_pair)
        s, low_count = remove_substring(s, low_priority_pair)

        return high_count * max(x, y) + low_count * min(x, y)

File: MaxScorebyremovingsubstring.cpp
class Solution {
public:
    int maximumGain(string s, int x, int y) {
        // Determine high and low priority pairs
        string highPriorityPair = x > y ? "ab" : "ba";
        string lowPriorityPair = x > y ? "ba" : "ab";
        int highValue = max(x, y);
        int lowValue = min(x, y);

        // First pass: remove high priority pairs and calculate score
        auto firstPassResult = removePairs(s, highPriorityPair);
        int totalScore = firstPassResult.second * highValue;

        // Second pass: remove low priority pairs and calculate score
        auto secondPassResult = removePairs(firstPassResult.first, lowPriorityPair);
        totalScore += secondPassResult.second * lowValue;

        return totalScore;
    }

private:
    pair<string, int> removePairs(const string& s, const string& pair) {
        string result;
        int count = 0;
        for (char c : s) {
            if (!result.empty() && result.back() == pair[0] && c == pair[1]) {
                result.pop_back();
                count++;
            } else {
                result.push_back(c);
            }
        }
        return {result, count};
    }
};
